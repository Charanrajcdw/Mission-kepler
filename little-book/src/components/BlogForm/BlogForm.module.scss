@import "../../scss/fonts";

.blogFormContainer {
  display: flex;
  flex-flow: column nowrap;
  height: 100%;
  color: var(--primary-font-bg);

  input,
  textarea {
    width: 100%;
    border: none;
    outline: none;
    font-family: $font-roboto;
    background-color: var(--modal-bg);
    color: var(--primary-font-bg);
  }

  input::placeholder,
  textarea::placeholder {
    color: var(--input-color);
  }

  .formTitle {
    font-family: $font-roboto;
    font-size: 1.25rem;
    font-weight: 400;
    margin-bottom: 28px;
  }

  .inputContainer {
    flex-grow: 1;

    .titleInput {
      margin: 20px 0;
      font-size: 1.7rem;
      font-weight: 400;
    }
    .detailsInput {
      margin: 30px 0;
      font-size: 1.05rem;
      resize: none;
    }
    &::-webkit-scrollbar {
      display: none;
    }
  }

  .buttonContainer {
    display: flex;
    justify-content: flex-end;
  }
}
