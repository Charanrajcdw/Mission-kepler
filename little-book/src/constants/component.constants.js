export const MODAL = {
  members: "Members",
  newBlog: "Add New Blog",
  remove: "Remove",
};

export const MEMBERS_LIST = {
  title: "Members",
  noMembers: "No members available",
};

export const BLOG_FORM = {
  title: "Add New Blog",
  button: "ADD",
  titlePlaceholder: "Name your blog",
  detailsPlaceholder: "Write Content Here ..",
  invalidValues: "Invalid title or description",
  invalidBlog: "Blog title already exists",
  blogAdded: "New Blog added successfully",
  localBlog: "local",
};

export const WARNING_MODAL = {
  title: "Unsaved changes will be lost. Are you sure to continue?",
  continue: "Continue",
  cancel: "Cancel",
};
