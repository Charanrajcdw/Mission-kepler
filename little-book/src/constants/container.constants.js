export const SIDEBAR = {
  logo: "Little Book",
  filter: "FILTER",
  noFilter: "Add blogs to show filters",
  viewMembers: "View Members",
  switchToDark: "Switch to Dark Mode",
  switchToLight: "Switch to Light Mode",
  blogs: "Blogs",
};
