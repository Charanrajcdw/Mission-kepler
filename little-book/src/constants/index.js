export * from "./url.constants";
export * from "./theme.constants";
