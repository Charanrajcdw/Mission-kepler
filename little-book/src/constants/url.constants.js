export const API_URLS = {
  blogs: "https://jsonmockserver.vercel.app/api/blogs",
  users: "https://jsonmockserver.vercel.app/api/users",
  images: "https://jsonmockserver.vercel.app/",
};
