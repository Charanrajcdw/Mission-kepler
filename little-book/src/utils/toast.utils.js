import { toast } from "react-toastify";

export const showSuccessToast = (message, currentTheme) => {
  toast.success(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};

export const showWarningToast = (message, currentTheme) => {
  toast.warn(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};
