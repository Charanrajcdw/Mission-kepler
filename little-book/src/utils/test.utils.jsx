import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import ThemeContextProvider from "../contexts/theme.context";
import blogSlice from "../store/slices/blogSlice";
import { THEME } from "../constants";

const customRender = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
    ...options
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return render(ui, { wrapper: ProviderWrapper, ...options });
};

const customCreateSnapshot = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return renderer.create(<ProviderWrapper>{ui}</ProviderWrapper>).toJSON();
};

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { customRender as render, customCreateSnapshot as createSnapshot };
