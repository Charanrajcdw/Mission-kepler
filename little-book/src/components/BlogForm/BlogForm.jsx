import { useContext, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import { ToastContainer } from "react-toastify";
import "react-toastify/dist/ReactToastify.min.css";
import PropTypes from "prop-types";
import styles from "./BlogForm.module.scss";
import Button from "../Button/Button";
import { ThemeContext } from "../../contexts/theme.context";
import { showSuccessToast, showWarningToast } from "../../utils/toast.utils";
import { BLOG_FORM } from "../../constants";
import { blogActions } from "../../store";

const BlogForm = ({ setCurrentModal }) => {
  const titleRef = useRef();
  const detailsRef = useRef();
  const { currentTheme } = useContext(ThemeContext);
  const { blogData } = useSelector((state) => state.blogs);
  const dispatch = useDispatch();

  const submitHandler = () => {
    const title = titleRef.current.value.trim();
    const details = detailsRef.current.value.trim();
    if (title.length <= 0 || details.length <= 0) {
      showWarningToast(BLOG_FORM.invalidValues, currentTheme);
    } else if (blogData.some((blog) => blog.title.toLowerCase() === title.toLowerCase())) {
      showWarningToast(BLOG_FORM.invalidBlog, currentTheme);
    } else {
      showSuccessToast(BLOG_FORM.blogAdded, currentTheme);
      dispatch(blogActions.addNewBlog({ title, details, type: BLOG_FORM.localBlog }));
      setCurrentModal();
    }
  };

  return (
    <div className={styles.blogFormContainer}>
      <ToastContainer />
      <p className={styles.formTitle}>{BLOG_FORM.title}</p>
      <div className={styles.inputContainer}>
        <input type="text" className={styles.titleInput} placeholder={BLOG_FORM.titlePlaceholder} ref={titleRef} />
        <textarea className={styles.detailsInput} placeholder={BLOG_FORM.detailsPlaceholder} ref={detailsRef} />
      </div>
      <div className={styles.buttonContainer}>
        <Button className="pinkBtn" clickHandler={submitHandler}>
          {BLOG_FORM.button}
        </Button>
      </div>
    </div>
  );
};

BlogForm.propTypes = {
  setCurrentModal: PropTypes.func.isRequired,
};

export default BlogForm;
