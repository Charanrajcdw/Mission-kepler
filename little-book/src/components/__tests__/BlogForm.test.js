import { render, screen, createSnapshot, userEvent } from "../../utils/test.utils.jsx";
import BlogForm from "../BlogForm/BlogForm";
import { BLOG_FORM } from "../../constants";

describe("BlogForm", () => {
  it("should render blog form", () => {
    render(<BlogForm modalHandler={() => {}} />);
    const titleInput = screen.getByPlaceholderText(BLOG_FORM.titlePlaceholder);
    const detailsInput = screen.getByPlaceholderText(BLOG_FORM.detailsPlaceholder);
    const submitButton = screen.getByRole("button");
    expect(titleInput).toBeInTheDocument();
    expect(detailsInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  it("should match blog form snapshot", () => {
    const blogForm = createSnapshot(<BlogForm modalHandler={() => {}} />);
    expect(blogForm).toMatchSnapshot();
  });

  it("should show invalid values warning toast", async () => {
    const user = userEvent.setup();
    render(<BlogForm modalHandler={() => {}} />);
    const submitButton = screen.getByRole("button");
    await user.click(submitButton);
    const warningToast = await screen.findByText(BLOG_FORM.invalidValues);
    expect(warningToast).toBeInTheDocument();
  });

  it("should show invalid title warning toast", async () => {
    const user = userEvent.setup();
    const initialBlogsData = {
      blogData: [{ title: "title1" }, { title: "title2" }],
    };
    render(<BlogForm modalHandler={() => {}} />, { preloadedState: { blogs: initialBlogsData } });
    const titleInput = screen.getByPlaceholderText(BLOG_FORM.titlePlaceholder);
    const detailsInput = screen.getByPlaceholderText(BLOG_FORM.detailsPlaceholder);
    const submitButton = screen.getByRole("button");
    await user.type(titleInput, "title1");
    await user.type(detailsInput, "details");
    await user.click(submitButton);
    const warningToast = await screen.findByText(BLOG_FORM.invalidBlog);
    expect(warningToast).toBeInTheDocument();
  });

  it("should show success messsage on submitting correct details", async () => {
    const user = userEvent.setup();
    render(<BlogForm modalHandler={() => {}} />);
    const titleInput = screen.getByPlaceholderText(BLOG_FORM.titlePlaceholder);
    const detailsInput = screen.getByPlaceholderText(BLOG_FORM.detailsPlaceholder);
    const submitButton = screen.getByRole("button");
    await user.type(titleInput, "blog title");
    await user.type(detailsInput, "blog details");
    await user.click(submitButton);
    const successToast = await screen.findByText(BLOG_FORM.blogAdded);
    expect(successToast).toBeInTheDocument();
  });
});
