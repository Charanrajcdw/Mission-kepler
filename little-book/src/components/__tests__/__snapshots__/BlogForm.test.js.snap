// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogForm should match blog form snapshot 1`] = `
<div
  className="blogFormContainer"
>
  <div
    className="Toastify"
  />
  <p
    className="formTitle"
  >
    Add New Blog
  </p>
  <div
    className="inputContainer"
  >
    <input
      className="titleInput"
      placeholder="Name your blog"
      type="text"
    />
    <textarea
      className="detailsInput"
      placeholder="Write Content Here .."
    />
  </div>
  <div
    className="buttonContainer"
  >
    <button
      className="btn pinkBtn"
      onClick={[Function]}
    >
      ADD
    </button>
  </div>
</div>
`;
